quic/s2n-quic-events/src/output/metrics.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{parser::File, Output};
+use proc_macro2::TokenStream;
+use quote::quote;
+
+mod aggregate;
+
+pub fn emit(output: &Output, files: &[File]) -> TokenStream {
+    let events = files
+        .iter()
+        .flat_map(|file| file.structs.iter())
+        .filter(|s| s.attrs.event_name.is_some())
+        .filter(|s| {
+            // Metrics is only connection-level events
+            s.attrs.subject.is_connection()
+        });
+
+    let mode = &output.mode;
+
+    let imports = output.mode.imports();
+    let receiver = output.mode.receiver();
+    let counter_increment = output.mode.counter_increment();
+    let counter_type = output.mode.counter_type();
+    let counter_init = output.mode.counter_init();
+    let counter_load = output.mode.counter_load();
+
+    let mut fields = quote!();
+    let mut init = quote!();
+    let mut record = quote!();
+    let mut subscriber = quote!();
+
+    for event in events {
+        let ident = &event.ident;
+        let counter = event.counter();
+        let snake = event.ident_snake();
+        let function = event.function();
+        let allow_deprecated = &event.attrs.allow_deprecated;
+
+        fields.extend(quote!(
+            #counter: #counter_type,
+        ));
+        init.extend(quote!(
+            #counter: #counter_init,
+        ));
+
+        record.extend(quote!(
+            self.recorder.increment_counter(#snake, self.#counter #counter_load as _);
```

Comment:
I *think* we'll want counter_swap(0) or something like that here -- I guess it depends on the metric system, some want each counter emission to have distinct counts (e.g., if you have +1, +1, +0, +1 and publishing every 60 seconds you'd emit =1, =1, =0, =1, not =1, =2, =2, =3). Others want the running sum (e.g., Prometheus largely expects a running sum).

Reply:
These metrics are connection level metrics that we have today, which are recorded once at the end of the connection. I just refactored the code to be in a single location rather than spread across everything. See https://github.com/aws/s2n-quic/blob/65d55a48ae4916c2a0c4ab9a90631668ba8ea23f/quic/s2n-quic-events/src/parser.rs#L358

Reply:
Ah, so swap or load is the same since we expect this to be terminal (i.e. no more increments afterwards).